File: algorithm/study/Core-09-Greedy/1744/python/src/solution.py
import sys
input = sys.stdin.readline

def main():
    N = int(input())
    pos = []   # 1보다 큰 수
    ones = 0
    neg = []   # 0보다 작은 수
    zeros = 0

    for _ in range(N):
        x = int(input())
        if x > 1:
            pos.append(x)
        elif x == 1:
            ones += 1
        elif x == 0:
            zeros += 1
        else:
            neg.append(x)

    total = ones  # 값 1은 그대로 더한다

    # 양수는 가장 큰 값부터 둘씩 묶는다
    pos.sort(reverse=True)
    for i in range(0, len(pos) - 1, 2):
        total += pos[i] * pos[i + 1]
    if len(pos) % 2 == 1:
        total += pos[-1]

    # 음수는 가장 작은 값부터 둘씩 묶는다
    neg.sort()
    for i in range(0, len(neg) - 1, 2):
        total += neg[i] * neg[i + 1]
    if len(neg) % 2 == 1:
        # 0이 있으면 남은 음수를 상쇄하고, 없으면 그대로 더한다
        if zeros == 0:
            total += neg[-1]

    print(total)

if __name__ == "__main__":
    main()

File: algorithm/study/Core-09-Greedy/1744/cpp/src/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N; cin >> N;
    vector<int> pos, neg;
    int ones = 0, zeros = 0;

    for(int i = 0; i < N; i++){
        int x; cin >> x;
        if(x > 1) pos.push_back(x);
        else if(x == 1) ones++;
        else if(x == 0) zeros++;
        else neg.push_back(x);
    }

    long long total = ones;

    sort(pos.rbegin(), pos.rend());
    for(int i = 0; i + 1 < (int)pos.size(); i += 2)
        total += (long long)pos[i] * pos[i+1];
    if(pos.size() % 2 == 1) total += pos.back();

    sort(neg.begin(), neg.end());
    for(int i = 0; i + 1 < (int)neg.size(); i += 2)
        total += (long long)neg[i] * neg[i+1];
    if(neg.size() % 2 == 1 && zeros == 0)
        total += neg.back();

    cout << total << "\n";
    return 0;
}
